### PR TITLE
feat: Apple Silicon Universal Support

### PR DESCRIPTION
**Pull request by lutzroeder**
#5475

@mmaietta

### CODE REVIEW

**Review by mmaietta — COMMENTED** [review]
With the introduction of `--universal` as an arch, I think this will break as node-gyp won't understand `universal`.
```
await this.installAppDependencies(platform, arch)
```
https://github.com/electron-userland/electron-builder/blob/master/packages/app-builder-lib/src/packager.ts#L430

It can process `--arch=all`, however, that doesn't work for me on Mac and it just produces the system's `process.arch`.
```
$ electron-builder node-gyp-rebuild --arch universal

Rebuild own native code

Options:
  --version   Show version number                                      [boolean]
  --help      Show help                                                [boolean]
  --platform  The target platform
                       [choices: "linux", "darwin", "win32"] [default: "darwin"]
  --arch      The target arch
             [choices: "ia32", "x64", "armv7l", "arm64", "all"] [default: "x64"]

Invalid values:
  Argument: arch, Given: "universal", Choices: "ia32", "x64", "armv7l", "arm64", "all"
error Command failed with exit code 1.
```

[UPDATE]
I'm starting to think we should just add a property specific to the builder config, like below. Partly because the packager logic is too tied to a specific architecture, and "universal" is technically not an architecture, but rather just combining/stitching assets into a fat binary.
```
    mac: {
        identity: <non-null>,
        universal: true,
        target: {
            target: 'dir',
            arch: [
                'arm64',
                'x64', 
            ]
        },
```
In taking this approach, the `target` field corresponds directly to an arch AND signs it. `universal` just validates that both arm64 and x64 are built, but doesn't force each individually to get signed, only the final fat app.

File: packages/app-builder-lib/src/macPackager.ts
Comment (lutzroeder) [reply packages/app-builder-lib/src/macPackager.ts:141]: That's not how this is being used in practice. Universal exists to avoid the existence of platform specific builds, not to bring them back in a hidden way or special case updaters. If an app builds universal, it would not build or ship x86 or arm64 targets.

Using a flag, where in the pipeline would the universal code run? It would ultimately have to do all the same things a regular MacPackager does and hide the existence of the x86 and arm64 build results. 

Is there a sample project or repro for the node-gyp-rebuild issue?

I can abandon this PR if you want to start a new one if there are clear answers but sounds like it isn't all that simple and more trading one set of problems for another?

**Comment by quanglam2807** [comment]
I have a better idea for this. Why don’t we introduce an option in build config. Something like `combine: true`. And if that is set, we will combine the arch-specific files together with lipo. 

I also agree with @mmaietta that universal binaries are only useful for initial downloads. For updates, `electron-updater` should just download arch-specific files.

If you go with this design, it will be much simpler as only need to add a hook to run after arch-specific builds are ready.

**Comment by mmaietta** [comment]
> If you go with this design, it will be much simpler as only need to add a hook to run after arch-specific builds are ready.

Perhaps a hook over here? https://github.com/electron-userland/electron-builder/blob/master/packages/app-builder-lib/src/packager.ts#L442
The only difficulty I'm running into is how to `map` `packager.pack(...) => Promise<string>` so that we return a target-arch-outputpath map/object of assets to lipo after (it doesn't return anything right now). Then it'd be easy to just pass that into the hook for `postProcessPackages(...)`

**Comment by lutzroeder** [comment]
@quanglam2807 @mmaietta to get on the same page on this, there aren't apps out there shipping a universal app and then upgrading to a platform specific version. Download and update any universal app, run an update and the result will be universal. With the "flag" approach the platform specific targets would need to be signed, notarized and published (on GitHub Releases) which would make CI pipelines more fragile and add to complexity (now there are 3 targets to choose from which we tried to avoid, users at different stages might rund a different version of the app). I'm not trying to argue which sets of pros and cons are preferable, just this is how it seems to be done by pretty much all app vendors so the goal is to optimize for this model.

**Comment by lutzroeder** [comment]
> Perhaps a hook over here? https://github.com/electron-userland/electron-builder/blob/master/packages/app-builder-lib/src/packager.ts#L442

What would this hook do? Create another MacPackager that doesn't build but can take inputs and `universal` architecture to create targets, then special case MacPackager so it doesn't sign or produce targets for `x64` and `arm64` if `universal` is set? Might be easier to have another PR to have a look?

**Comment by mmaietta** [comment]
Thanks so much for tackling this @lutzroeder! The work is fantastic and covers almost all the use cases. The main issue lies with native modules as `installAppDependencies` must be called prior to any packing and per-architecture. `universal` unfortunately isn't accepted by node-gyp itself (`all` is though). I think we shouldn't move `installAppDependencies` into `doPack` as it convolutes the build stage's description of just "packing" (I suppose a simple rename could change that explicit description though, perhaps `buildPackage`?)
I think moving `installAppDependencies` here might solve the issue such that `doPack` always installs/rebuilds the correct arch's native addons: https://github.com/electron-userland/electron-builder/blob/70708a470ba148787c110be37c482e12c52e9351/packages/app-builder-lib/src/platformPackager.ts#L161-L162
Haven't tested that locally yet, will need to stash my current investigations.

File: packages/app-builder-lib/src/macPackager.ts
Comment (mmaietta) [reply packages/app-builder-lib/src/macPackager.ts:125]: Sorry for the delay on this one.
You're completely correct that the approachs will be trading one set of problems for another. 😅 

Quick note: `electron-updater` currently expects to find an arch-specific build. 
https://github.com/electron-userland/electron-builder/blob/2b116d138e9effa3314fe6ff4474b639f7fddf47/packages/electron-updater/src/MacUpdater.ts#L34-L35
After much time for users to upgrade to M1, the universal binaries will be dropped anyhow, and builds will just need to include arm64, which electron-updater will already be ready to handle.

**Comment by lutzroeder** [comment]
@mmaietta updated to call `installAppDependencies` from `doPack`:

```
~/electron-quick-start: npx electron-builder --mac --universal
  • electron-builder  version=22.10.0 os=20.2.0
  • writing effective config  file=dist/builder-effective-config.yaml
  • rebuilding native dependencies  dependencies=leveldown@5.6.0 platform=darwin arch=x64
  • packaging       platform=darwin arch=x64 electron=11.1.0 appOutDir=dist/mac-universal-x64
  • default Electron icon is used  reason=application icon is not set
  • rebuilding native dependencies  dependencies=leveldown@5.6.0 platform=darwin arch=arm64
  • packaging       platform=darwin arch=arm64 electron=11.1.0 appOutDir=dist/mac-universal-arm64
  • packaging       platform=darwin arch=universal electron=11.1.0 appOutDir=dist/mac-universal
```

**Comment by mmaietta** [comment]
Nice! Those logs look perfect.

For the interim, I think this is a good way to approach it without very large changes for a single hook. A larger refactor without an explicit need for a post-all-packed hook is just asking for bugs to be created. I'm still wary of the "universal" as its own arch as anyone building all 3 archs (x64, arm64, universal) with native addons rebuilding may take a very long time. Oh well 😅 

I gave it a whirl locally and it works perfect on the first run. lipo results:
```
$ find dist/artifacts build -name '*.node' | xargs -I {} lipo -info "{}"
Architectures in the fat file: dist/artifacts/local/mac-universal/Electron.app/Contents/MacOS/app.node are: x86_64 arm64 
```
It does throw an error on subsequent builds that is an easy fix:
```
  • packaging       platform=darwin arch=arm64 electron=11.0.3 appOutDir=dist/artifacts/local/mac-universal-arm64
  • packaging       platform=darwin arch=universal electron=11.0.3 appOutDir=dist/artifacts/local/mac-universal
  ⨯ The out path "/shell-builder/dist/artifacts/local/mac-universal/Electron.app" already exists and force is not set to true 
```
Just need to add `force: true` to the `makeUniversalApp` config to enable overwriting

The only remaining step AFAICT is to get this working with the electron-builder config so that this works:
```
    mac: {
        target: {
            target: 'dir',
            arch: 'universal'
        },
```

**Comment by lutzroeder** [comment]
> The only remaining step AFAICT is to get this working with the electron-builder config so that this works

The config gives `Invalid configuration object. electron-builder 22.10.0 has been initialized using a configuration object that does not match the API schema.`?

**Comment by mmaietta** [comment]
Yep! You'll need to update the scheme.json. Just gotta add `"universal",` here:
https://github.com/electron-userland/electron-builder/blob/master/packages/app-builder-lib/scheme.json#L4955-L4980

Now, how can we fix that test?

**Comment by lutzroeder** [comment]
> Now, how can we fix that test?

That should probably be a separate PR as it happens for blank PRs like #5488 as well. We should fix the test first and merge vs. including it in this PR which has higher risk to be reverted. Can you have a look?

**Comment by mmaietta** [comment]
Your tests pass here!
But locally they fail for me, I get:
```
Snapshot name: `two-package 1`

    - Snapshot  - 3
    + Received  + 3

    @@ -68,21 +68,21 @@
                  "url": "TestApp-1.1.0-mac-universal.zip",
                },
                Object {
                  "sha512": "@sha512",
                  "size": "@size",
    -             "url": "TestApp-1.1.0-mac.dmg",
    +             "url": "TestApp-1.1.0-mac-universal.dmg",
                },
                Object {
                  "sha512": "@sha512",
                  "size": "@size",
    -             "url": "TestApp-1.1.0-mac-arm64.dmg",
    +             "url": "TestApp-1.1.0-mac.dmg",
                },
                Object {
                  "sha512": "@sha512",
                  "size": "@size",
    -             "url": "TestApp-1.1.0-mac-universal.dmg",
    +             "url": "TestApp-1.1.0-mac-arm64.dmg",
                },
              ],
              "path": "TestApp-1.1.0-mac.zip",
              "releaseDate": "@releaseDate",
              "sha512": "@sha512",

      235 |   }
      236 | 
    > 237 |   expect(objectToCompare).toMatchSnapshot()
```

I feel like something is wrong with the build node as these results are way too inconsistent.
Yours fllows this order from packages/electron-builder/src/builder.ts `[Arch.x64, Arch.arm64, Arch.universal]`, whereas mine is in some weird format.

**Comment by lutzroeder** [comment]
@mmaietta not sure how this explains why #5488 is failing? This PR doesn't reorder `x64` and `arm64` and other PRs are failing as well.

**Comment by quanglam2807** [comment]
Thank you so much, @lutzroeder. I'm so excited for this. I just realize Mac App Store only accepts a single `pkg` file, so universal binary is a must. Hope this PR can be merged soon so we can then test in production.

File: packages/app-builder-lib/src/packager.ts
Comment (mmaietta) [inline packages/app-builder-lib/src/packager.ts:474]: Minor: Since `installAppDependencies` is isolated to `doPack` now, can we move this to `platformPackager` and use `protected` instead of `public`?

**Comment by lutzroeder** [comment]
@mmaietta @develar can we merge this PR and #5488 and for another round of prerelease testing?

File: packages/app-builder-lib/src/macPackager.ts
Comment (mmaietta) [inline packages/app-builder-lib/src/macPackager.ts:125]: Minor and probably pedantic, but if we were to consolidate this logic into a `map` it might be less lengthy of a `doPack` override.
```
const appFile = `${this.appInfo.productFilename}.app`;
const [x64AppPath, arm64AppPath] = [Arch.x64, Arch.arm64].map(async arch => {
  const destDir = appOutDir + '-' + Arch[arch];
  await super.doPack(outDir, destDir, platformName, arch, platformSpecificBuildOptions, targets, false);
  return path.join(destDir, appFile)
})
const { version, name } = this.info.framework
log.info({
  platform: platformName,
  arch: Arch[arch],
  [name]: version,
  appOutDir: log.filePath(appOutDir),
}, `packaging`)
const { makeUniversalApp } = require('@electron/universal');
await makeUniversalApp({ x64AppPath, arm64AppPath, outAppPath: path.join(appOutDir, appFile), force: true });

const rmdir = promisify(require('fs').rmdir);
archOutput.forEach(async dir => { await rmdir(dir, { recursive: true }) });
await this.doSignAfterPack(outDir, appOutDir, platformName, arch, platformSpecificBuildOptions, targets);
```

Regardless, this is fantastic work! Everything checks out locally for me as well.

**Review by mmaietta — APPROVED** [review]
Fantastic work @lutzroeder! 🎉

File: packages/app-builder-lib/src/packager.ts
Comment (lutzroeder) [reply packages/app-builder-lib/src/packager.ts:474]: `installAppDependencies` has a bunch of dependencies on `Packager` and not sure it should be moved.  If at all, this probably should be a separate refactoring.

**Comment by mmaietta** [comment]
I think we should also merge these for the prerelease cut:
https://github.com/electron-userland/electron-builder/pull/5488 https://github.com/electron-userland/electron-builder/pull/5489 https://github.com/electron-userland/electron-builder/pull/5484
CC @quanglam2807 

@lutzroeder I, unfortunately, don't have merge permissions for this project. It'd be awesome if I did though!

**Comment by mmaietta** [comment]
Quick follow up, can you please update this PR's `yarn.lock` 🙂 

Also, @lutzroeder I've pushed a "PreRelease" branch to my fork containing multiple fixes from open PRs we've worked on (dmg, universal, signing, unit tests). Feel free to generate patch files from it to use for the interim.
https://github.com/mmaietta/electron-builder/tree/PreRelease

Just compile, link, and use rsync to migrate all the files so you can run `patch-package` on your project.
Handy cmd: `rsync -avh --exclude=package.json --exclude=yalc.sig .yalc/ node_modules/`
You'll need to manually `yarn add -D istextorbinary @electron/universal` in your project for now because `yarn install` can't utilize a patched package.json to update the yarn.lock.

**Comment by develar** [comment]
Thanks for your work, you are awesome :)

I cannot merge https://github.com/electron-userland/electron-builder/pull/5488 since not clear for me, is it safe just revert or it is possible just add required patterns to auto sign.

**Comment by lutzroeder** [comment]
@develar note this PR is orthogonal to #5488. #5488 tracks an existing regression, this is a new feature.

**Comment by lutzroeder** [comment]
> Quick follow up, can you please update this PR's yarn.lock

@mmaietta ran `yarn install` and added the updated `yarn.lock`. Is that all that is needed for this PR?

**Review by mmaietta — APPROVED** [review]
LGTM

**Comment by lutzroeder** [comment]
@develar @mmaietta what does it take to push a new prerelease with the recent changes?

Is there a doc page how to build `electron-builder` locally and fully patch the `node_modules` of a project?

**Comment by mmaietta** [comment]
I have no idea in terms of a release 😅 

In terms of patch package for tracking latest master, here's how I've done it:
Step 1
```
yarn compile && \
(cd ./packages/app-builder-lib && yalc push) && \
(cd ./packages/builder-util && yalc push) && \
(cd ./packages/builder-util-runtime && yalc push) && \
(cd ./packages/dmg-builder && yalc push) && \
(cd ./packages/electron-builder && yalc push) && \
(cd ./packages/electron-publish && yalc push) && \
(cd ./packages/electron-builder-squirrel-windows && yalc push) && \
(cd ./packages/electron-forge-maker-appimage && yalc push) && \
(cd ./packages/electron-forge-maker-nsis && yalc push) && \
(cd ./packages/electron-forge-maker-nsis-web && yalc push) && \
(cd ./packages/electron-forge-maker-snap && yalc push) && \
(cd ./packages/electron-updater && yalc push)
```
Step 2 in your project:
```
yalc remove --all
yalc retreat --all
yarn install --check-files --force --ignore-scripts
yalc link app-builder-lib builder-util builder-util-runtime dmg-builder electron-builder electron-publish electron-builder-squirrel-windows electron-forge-maker-appimage electron-forge-maker-nsis electron-forge-maker-nsis-web electron-forge-maker-snap electron-updater
rsync -avh --exclude=package.json --exclude=yalc.sig .yalc/ node_modules/
yarn patch-package app-builder-lib --exclude
yarn patch-package builder-util --exclude
yarn patch-package electron-builder --exclude
yarn patch-package electron-updater --exclude
```
If you hit a permission issue during `yarn install`, then just rerun the `retreat` and `remove` commands until it properly resets the permissions in node_modules. Note: the `--ignore-scripts` is necessary to avoid applying your current patch files as it'll break the subsequent attempts to create new patches.

**Comment by develar** [comment]
`22.10.4` is released as pre-release.

**Comment by mmaietta** [comment]
@lutzroeder EB provides an `electronDist` param for custom electron dists/forks, which is what my team uses and something I think we may have overlooked. Not a bug as part of this PR as this issue already exists for compiling 2+ archs on a single build run, as there will always be only one path to an arch-specific electron dist.

For building universal, we need to provide 2 different paths, one for each arch. However, `electronDist` only accepts one path to a corresponding `.app`.
What are your thoughts on us enabling `electronDist` to accept a dictionary of `${platform}-${arch} : <path>` key-value pairs? It'd retain backward compatibility while enabling custom forks for universal builds.

CC @quanglam2807

**Comment by thaddeusm** [comment]
Has anyone else tried publishing these files to Github along with other targets?

I am able to build locally (22.10.4) but publishing does not trigger an upload of Mac builds, though Windows assets are being uploaded as usual for my configuration.

**Comment by quanglam2807** [comment]
@mmaietta Instead of using dictionary, wouldn't it better to allow specifying `electronDist` as a function? I think that'd be more flexible.

**Comment by quanglam2807** [comment]
@lutzroeder I just gave this a test in production. 

`@electron/universal` validates SHA of all non-binary files and `app.asar` to make sure that they are consistent between arm64/x64 builds:
https://github.com/electron/universal/blob/a05a5e6db87e88ffa60038de50a5ec930b8f64be/src/index.ts#L108

`Contents/Info.plist` SHAs are also inconsistent because `electron-builder`.
- adds `AsarIntegrity` field  (SHA of `app.asar`) to `Contents/Info.plist`.
- SHA of `Contents/Info.plist` is inconsistent between builds (we generate them separately for each arch-specific build).
This

**Comment by quanglam2807** [comment]
I just checked again, it's ok for SHAs of `app.asar` to be different. So I think we need to strip `AsarIntegrity` field from `Info.plist`.

**Comment by quanglam2807** [comment]
We can add a conditional check here: https://github.com/electron-userland/electron-builder/blob/ca20151c3416324d2413f451dea0c9e3853bab79/packages/app-builder-lib/src/platformPackager.ts#L229 but I'm not sure if removing `AsarIntegrity` would do any harms.

**Comment by lutzroeder** [comment]
@quanglam2807 what is causing them to be different?

**Comment by quanglam2807** [comment]
> @quanglam2807 what is causing them to be different?

We generate app.asar for each arch-specific build separately so they're different files with different SHAs.

**Comment by lutzroeder** [comment]
@quanglam2807 do the `app.asar` files contain arch-specific binaries or do they contain something else that would make them different? Is the difference is arch-specific binaries not sure how that would work as `lipo` is expected to merge such binaries.

**Comment by lutzroeder** [comment]
Follow-up #5495

**Comment by quanglam2807** [comment]
> @quanglam2807 do the `app.asar` files contain arch-specific binaries or do they contain something else that would make them different? If the difference is arch-specific binaries not sure how that would work as `lipo` is expected to merge such binaries.

`app.asar` normally doesn't contain binaries. If it does, `electron-builder` will automatically unpack them from the asar file.

We generate x64 & arm64 builds separately before merging them together. So for each arch, we generate different app.asar file. These files are identical in contents, but could have different SHA because they differ in metadata (eg date created).

So in this case, we can opt to remove AsarIntegrity from Info.plist. Or we can unify the app.asar process for universal build.

**Comment by lutzroeder** [comment]
@quanglam2807 do you have a repro or is there some extra logging that needs to be enabled? I've been running with 20.10.4, pushed a release and haven't seen this error.

**Comment by quanglam2807** [comment]
@lutzroeder You can clone this repo: https://github.com/webcatalog/translatium

Then remove the patch: `patches/app-builder-lib+22.10.4.patch`. Run `yarn` and then `yarn dist-dev`.

File: packages/app-builder-lib/src/macPackager.ts
Comment (vadim-termius) [inline packages/app-builder-lib/src/macPackager.ts:98]: Hi @lutzroeder, 

Thank you for the universal support.

Why do you overwrite `platformName` and `platformSpecificBuildOptions` instead of passing arguments directly? It breaks mas builds.

Comment (lutzroeder) [reply packages/app-builder-lib/src/macPackager.ts:98]: Good catch, these should pass through. Can you create an issue to track and/or feel free to send a PR.

**Comment by fabiospampinato** [comment]
Does this PR take into account `electron-updater`? 

Because when I switch from the `x64` to the `universal` architecture I'm not getting the build artifacts that I would need, just a single file at `mac-universal/MyApp.app` and this error in the terminal:

```
Error: spawn zip ENOENT
    at Process.ChildProcess._handle.onexit (internal/child_process.js:264:19)
    at onErrorNT (internal/child_process.js:456:16)
    at processTicksAndRejections (internal/process/task_queues.js:80:21)
```

I'm under Mojave.

**Comment by lutzroeder** [comment]
@fabiospampinato see #5504 which is not included in a preview build yet and #5495.

**Comment by fabiospampinato** [comment]
@lutzroeder Would that probably address my issue? 🤔 I can get a universal build of the app, but I can't get the dmg and accessory files. Seeing as an error, possibly related to packing the app, got thrown at build time I don't think that would address my issue.

**Pull request merged**
